pyembed/src/importer.rs

```diff
@@ -900,7 +904,219 @@ impl OxidizedFinder {
+        // PyImport_GetImporter checking for the existence of config->run_filename
+        // has been setup, so it's safe to import os now.
+        let paths = PyList::new(py, &[py.import("os")?.call(py, "fspath", (path,), None)?]);
+        _PathEntryFinder::create_instance(py, finder.into_object(), paths, pkg)
+    }
+}
+
+// A (mostly compliant) `importlib.abc.PathEntryFinder` that delegates paths
+// within the current executable to the `OxidizedFinder` whose `path_hook`
+// method created it. This is a private implementation detail.
+py_class!(class _PathEntryFinder |py| {
+    // A `importlib.abc.MetaPathFinder`, presumably but not necessarily an
+    // `OxidizedFinder`. However, `_PathEntryFinder::iter_modules` will raise
+    // a `TypeError` when called if `finder` is not an `OxidizedFinder`.
+    data finder: PyObject;
+    // A list containing a single either str or bytes, normalized from the `path`
+    // argument to `OxidizedFinder::path_hook`.
+    data path: PyList;
+    // If the entry of `path` is `os.path.join(sys.executable, "pkg", "mod")`,
+    // then `package` is `"pkg.mod"`.
+    data package: String;
+
+    def find_spec(&self, fullname: &str, target: Option<PyModule> = None) -> PyResult<Option<PyObject>> {
+        self.find_spec_impl(py, fullname, target)
+    }
+
+    def invalidate_caches(&self) -> PyResult<PyObject> {
+        self.finder(py).call_method(py, "invalidate_caches", NoArgs, None)
+    }
+
+    def iter_modules(&self, prefix: &str = "") -> PyResult<PyList> {
+        self.iter_modules_impl(py, prefix)
+    }
+
+    // Private getter. Just for testing.
+    @property def _package(&self) -> PyResult<String> {
+        Ok(self.package(py).clone())
+    }
+});
+
+impl _PathEntryFinder {
+    // `name` is considered _visible_ in `package` if `name` is in the first
+    // level of the package. The computation is purely textual. See this
+    // module's `test_path_entry_finder::test_is_visible` test.
+    fn is_visible(package: &str, name: &str) -> bool {
+        if package.starts_with('.') || package.ends_with('.') || package.contains("..") {
+            return false;
+        }
+        // This implementation can be simplified when str::rsplit_once stabilize
+        // https://doc.rust-lang.org/std/primitive.str.html#method.rsplit_once
+        let mut split = name.rsplitn(2, '.');
+        match split.next() {
+            Some(module) if !module.is_empty() && !module.contains('.') => {}
+            _ => {
+                return false;
+            }
+        }
+        match split.next() {
+            Some(pkg) if !package.is_empty() && pkg == package => true,
+            None if package.is_empty() => true,
+            _ => false,
+        }
+    }
+
+    fn find_spec_impl(
+        &self,
+        py: Python,
+        fullname: &str,
+        target: Option<PyModule>,
+    ) -> PyResult<Option<PyObject>> {
+        if !_PathEntryFinder::is_visible(self.package(py), fullname) {
```

Comment:
Can we use `Self::is_visible()` here to avoid typing the type name?

Reply:
3da60fe800f5e28a2b5493fe99de5622473852b0